File: test-automation.js
#!/usr/bin/env node

/**
 * Test script for Open Collaboration Tools Automation API
 * Usage: node test-automation.js [create|join|both]
 */

const http = require('http');

const PORT = 8443;
const BASE_URL = `http://127.0.0.1:${PORT}`;

// ANSI color codes
const colors = {
    reset: '\x1b[0m',
    green: '\x1b[32m',
    red: '\x1b[31m',
    blue: '\x1b[34m',
    yellow: '\x1b[33m'
};

// Helper function to make HTTP requests
function makeRequest(data) {
    return new Promise((resolve, reject) => {
        const postData = JSON.stringify(data);
        
        const options = {
            hostname: '127.0.0.1',
            port: PORT,
            path: '/',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(postData)
            }
        };

        const req = http.request(options, (res) => {
            let responseData = '';

            res.on('data', (chunk) => {
                responseData += chunk;
            });

            res.on('end', () => {
                try {
                    const result = JSON.parse(responseData);
                    resolve({ statusCode: res.statusCode, data: result });
                } catch (e) {
                    resolve({ statusCode: res.statusCode, data: responseData });
                }
            });
        });

        req.on('error', (e) => {
            reject(e);
        });

        req.write(postData);
        req.end();
    });
}

// Test: Create session
async function testCreate() {
    console.log(`${colors.blue}[TEST 1]${colors.reset} Creating new session...`);
    
    try {
        const result = await makeRequest({ action: 'create' });
        
        console.log(`HTTP Status: ${result.statusCode}`);
        console.log(`Response: ${JSON.stringify(result.data, null, 2)}`);
        
        if (result.statusCode === 200 && result.data.success) {
            console.log(`${colors.green}✓ SUCCESS${colors.reset} - Session created!`);
            console.log(`Room ID: ${colors.green}${result.data.roomId}${colors.reset}`);
            console.log('');
            return result.data.roomId;
        } else {
            console.log(`${colors.red}✗ FAILED${colors.reset} - Could not create session`);
            console.log('');
            return null;
        }
    } catch (error) {
        console.log(`${colors.red}✗ ERROR${colors.reset} - ${error.message}`);
        console.log('');
        return null;
    }
}

// Test: Join session
async function testJoin(roomId) {
    if (!roomId) {
        console.log(`${colors.red}Error: No room ID provided${colors.reset}`);
        return;
    }
    
    console.log(`${colors.blue}[TEST 2]${colors.reset} Joining session: ${roomId}`);
    
    try {
        const result = await makeRequest({ 
            action: 'join', 
            roomId: roomId 
        });
        
        console.log(`HTTP Status: ${result.statusCode}`);
        console.log(`Response: ${JSON.stringify(result.data, null, 2)}`);
        
        if (result.statusCode === 200 && result.data.success) {
            console.log(`${colors.green}✓ SUCCESS${colors.reset} - Joined session!`);
            console.log('');
        } else {
            console.log(`${colors.red}✗ FAILED${colors.reset} - Could not join session`);
            console.log('');
        }
    } catch (error) {
        console.log(`${colors.red}✗ ERROR${colors.reset} - ${error.message}`);
        console.log('');
    }
}

// Test: Invalid request
async function testInvalid() {
    console.log(`${colors.blue}[TEST 3]${colors.reset} Testing invalid request (should fail)...`);
    
    try {
        const result = await makeRequest({ action: 'invalid' });
        
        console.log(`HTTP Status: ${result.statusCode}`);
        console.log(`Response: ${JSON.stringify(result.data, null, 2)}`);
        
        if (result.statusCode === 400) {
            console.log(`${colors.green}✓ SUCCESS${colors.reset} - Error handling works correctly`);
            console.log('');
        } else {
            console.log(`${colors.red}✗ UNEXPECTED${colors.reset} - Expected 400 status code`);
            console.log('');
        }
    } catch (error) {
        console.log(`${colors.red}✗ ERROR${colors.reset} - ${error.message}`);
        console.log('');
    }
}

// Test: Missing roomId for join
async function testMissingRoomId() {
    console.log(`${colors.blue}[TEST 4]${colors.reset} Testing join without roomId (should fail)...`);
    
    try {
        const result = await makeRequest({ action: 'join' });
        
        console.log(`HTTP Status: ${result.statusCode}`);
        console.log(`Response: ${JSON.stringify(result.data, null, 2)}`);
        
        if (result.statusCode === 400 && result.data.error) {
            console.log(`${colors.green}✓ SUCCESS${colors.reset} - Validation works correctly`);
            console.log('');
        } else {
            console.log(`${colors.red}✗ UNEXPECTED${colors.reset} - Expected 400 with error`);
            console.log('');
        }
    } catch (error) {
        console.log(`${colors.red}✗ ERROR${colors.reset} - ${error.message}`);
        console.log('');
    }
}

// Check if service is running
async function checkService() {
    console.log('Checking if automation service is running...');
    
    try {
        const result = await makeRequest({ action: 'create' });
        console.log(`${colors.green}✓ Service is running on port ${PORT}${colors.reset}`);
        console.log('');
        return true;
    } catch (error) {
        if (error.code === 'ECONNREFUSED') {
            console.log(`${colors.red}✗ Service is NOT running on port ${PORT}${colors.reset}`);
            console.log('');
            console.log('Please ensure:');
            console.log('1. VSCode is running');
            console.log('2. Open Collaboration Tools extension is installed');
            console.log('3. Extension is activated');
            console.log('');
            return false;
        }
        throw error;
    }
}

// Main function
async function main() {
    console.log('================================================');
    console.log('  OCT Automation API Test Script');
    console.log('================================================');
    console.log('');
    
    const serviceRunning = await checkService();
    if (!serviceRunning) {
        process.exit(1);
    }
    
    const command = process.argv[2] || 'both';
    const roomIdArg = process.argv[3];
    
    switch (command) {
        case 'create':
            await testCreate();
            break;
            
        case 'join':
            if (!roomIdArg) {
                console.log(`${colors.red}Error: Room ID required${colors.reset}`);
                console.log('Usage: node test-automation.js join <room-id>');
                process.exit(1);
            }
            await testJoin(roomIdArg);
            break;
            
        case 'invalid':
            await testInvalid();
            break;
            
        case 'validation':
            await testMissingRoomId();
            break;
            
        case 'both':
            const roomId = await testCreate();
            if (roomId) {
                console.log('Waiting 2 seconds before join test...');
                await new Promise(resolve => setTimeout(resolve, 2000));
                await testJoin(roomId);
            }
            await testInvalid();
            await testMissingRoomId();
            break;
            
        case 'all':
            const rid = await testCreate();
            if (rid) {
                console.log('Waiting 2 seconds before join test...');
                await new Promise(resolve => setTimeout(resolve, 2000));
                await testJoin(rid);
            }
            await testInvalid();
            await testMissingRoomId();
            break;
            
        default:
            console.log('Usage: node test-automation.js [create|join <room-id>|both|all|invalid|validation]');
            console.log('');
            console.log('Examples:');
            console.log('  node test-automation.js create              # Test create session');
            console.log('  node test-automation.js join abc123         # Test join session');
            console.log('  node test-automation.js both                # Test create and join');
            console.log('  node test-automation.js all                 # Run all tests');
            console.log('  node test-automation.js invalid             # Test error handling');
            console.log('  node test-automation.js validation          # Test validation');
            process.exit(1);
    }
    
    console.log('================================================');
    console.log('  Test Complete');
    console.log('================================================');
}

// Run if called directly
if (require.main === module) {
    main().catch(error => {
        console.error(`${colors.red}Fatal error:${colors.reset}`, error);
        process.exit(1);
    });
}

module.exports = { makeRequest, testCreate, testJoin };

File: test-automation.py
#!/usr/bin/env python3

"""
Test script for Open Collaboration Tools Automation API
Usage: python test-automation.py [create|join|both]
"""

import sys
import json
import time
import requests
from typing import Optional, Dict, Any

# Configuration
PORT = 8443
BASE_URL = f"http://127.0.0.1:{PORT}"

# ANSI color codes
class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    BLUE = '\033[94m'
    YELLOW = '\033[93m'
    RESET = '\033[0m'

def make_request(data: Dict[str, Any]) -> tuple[int, Dict[str, Any]]:
    """Make HTTP POST request to automation API"""
    try:
        response = requests.post(
            BASE_URL,
            json=data,
            headers={'Content-Type': 'application/json'}
        )
        return response.status_code, response.json()
    except requests.exceptions.ConnectionError:
        print(f"{Colors.RED}✗ Cannot connect to service{Colors.RESET}")
        print(f"\nPlease ensure:")
        print("1. VSCode is running")
        print("2. Open Collaboration Tools extension is installed")
        print("3. Extension is activated")
        sys.exit(1)
    except Exception as e:
        print(f"{Colors.RED}✗ Error: {str(e)}{Colors.RESET}")
        sys.exit(1)

def test_create() -> Optional[str]:
    """Test create session"""
    print(f"{Colors.BLUE}[TEST 1]{Colors.RESET} Creating new session...")
    
    status_code, data = make_request({'action': 'create'})
    
    print(f"HTTP Status: {status_code}")
    print(f"Response: {json.dumps(data, indent=2)}")
    
    if status_code == 200 and data.get('success'):
        room_id = data.get('roomId')
        print(f"{Colors.GREEN}✓ SUCCESS{Colors.RESET} - Session created!")
        print(f"Room ID: {Colors.GREEN}{room_id}{Colors.RESET}")
        print()
        return room_id
    else:
        print(f"{Colors.RED}✗ FAILED{Colors.RESET} - Could not create session")
        print()
        return None

def test_join(room_id: str) -> bool:
    """Test join session"""
    if not room_id:
        print(f"{Colors.RED}Error: No room ID provided{Colors.RESET}")
        return False
    
    print(f"{Colors.BLUE}[TEST 2]{Colors.RESET} Joining session: {room_id}")
    
    status_code, data = make_request({
        'action': 'join',
        'roomId': room_id
    })
    
    print(f"HTTP Status: {status_code}")
    print(f"Response: {json.dumps(data, indent=2)}")
    
    if status_code == 200 and data.get('success'):
        print(f"{Colors.GREEN}✓ SUCCESS{Colors.RESET} - Joined session!")
        print()
        return True
    else:
        print(f"{Colors.RED}✗ FAILED{Colors.RESET} - Could not join session")
        print()
        return False

def test_invalid():
    """Test invalid request"""
    print(f"{Colors.BLUE}[TEST 3]{Colors.RESET} Testing invalid request (should fail)...")
    
    status_code, data = make_request({'action': 'invalid'})
    
    print(f"HTTP Status: {status_code}")
    print(f"Response: {json.dumps(data, indent=2)}")
    
    if status_code == 400:
        print(f"{Colors.GREEN}✓ SUCCESS{Colors.RESET} - Error handling works correctly")
        print()
    else:
        print(f"{Colors.RED}✗ UNEXPECTED{Colors.RESET} - Expected 400 status code")
        print()

def test_missing_room_id():
    """Test join without roomId"""
    print(f"{Colors.BLUE}[TEST 4]{Colors.RESET} Testing join without roomId (should fail)...")
    
    status_code, data = make_request({'action': 'join'})
    
    print(f"HTTP Status: {status_code}")
    print(f"Response: {json.dumps(data, indent=2)}")
    
    if status_code == 400 and 'error' in data:
        print(f"{Colors.GREEN}✓ SUCCESS{Colors.RESET} - Validation works correctly")
        print()
    else:
        print(f"{Colors.RED}✗ UNEXPECTED{Colors.RESET} - Expected 400 with error")
        print()

def test_custom_server():
    """Test with custom server URL"""
    print(f"{Colors.BLUE}[TEST 5]{Colors.RESET} Testing with custom server URL...")
    
    status_code, data = make_request({
        'action': 'create',
        'serverUrl': 'https://api.open-collab.tools/'
    })
    
    print(f"HTTP Status: {status_code}")
    print(f"Response: {json.dumps(data, indent=2)}")
    
    if status_code == 200 and data.get('success'):
        print(f"{Colors.GREEN}✓ SUCCESS{Colors.RESET} - Custom server URL works!")
        print()
    else:
        print(f"{Colors.RED}✗ FAILED{Colors.RESET} - Custom server URL failed")
        print()

def check_service():
    """Check if service is running"""
    print("Checking if automation service is running...")
    try:
        response = requests.post(
            BASE_URL,
            json={'action': 'create'},
            timeout=2
        )
        print(f"{Colors.GREEN}✓ Service is running on port {PORT}{Colors.RESET}")
        print()
        return True
    except requests.exceptions.ConnectionError:
        print(f"{Colors.RED}✗ Service is NOT running on port {PORT}{Colors.RESET}")
        print()
        return False
    except Exception as e:
        print(f"{Colors.RED}✗ Error: {str(e)}{Colors.RESET}")
        print()
        return False

def main():
    """Main function"""
    print("=" * 50)
    print("  OCT Automation API Test Script (Python)")
    print("=" * 50)
    print()
    
    if not check_service():
        print("\nPlease ensure:")
        print("1. VSCode is running")
        print("2. Open Collaboration Tools extension is installed")
        print("3. Extension is activated")
        sys.exit(1)
    
    command = sys.argv[1] if len(sys.argv) > 1 else 'both'
    room_id_arg = sys.argv[2] if len(sys.argv) > 2 else None
    
    if command == 'create':
        test_create()
    
    elif command == 'join':
        if not room_id_arg:
            print(f"{Colors.RED}Error: Room ID required{Colors.RESET}")
            print("Usage: python test-automation.py join <room-id>")
            sys.exit(1)
        test_join(room_id_arg)
    
    elif command == 'invalid':
        test_invalid()
    
    elif command == 'validation':
        test_missing_room_id()
    
    elif command == 'server':
        test_custom_server()
    
    elif command == 'both':
        room_id = test_create()
        if room_id:
            print("Waiting 2 seconds before join test...")
            time.sleep(2)
            test_join(room_id)
        test_invalid()
        test_missing_room_id()
    
    elif command == 'all':
        room_id = test_create()
        if room_id:
            print("Waiting 2 seconds before join test...")
            time.sleep(2)
            test_join(room_id)
        test_invalid()
        test_missing_room_id()
        test_custom_server()
    
    else:
        print("Usage: python test-automation.py [create|join <room-id>|both|all|invalid|validation|server]")
        print()
        print("Examples:")
        print("  python test-automation.py create              # Test create session")
        print("  python test-automation.py join abc123         # Test join session")
        print("  python test-automation.py both                # Test create and join")
        print("  python test-automation.py all                 # Run all tests")
        print("  python test-automation.py invalid             # Test error handling")
        print("  python test-automation.py validation          # Test validation")
        print("  python test-automation.py server              # Test custom server")
        sys.exit(1)
    
    print("=" * 50)
    print("  Test Complete")
    print("=" * 50)

if __name__ == '__main__':
    main()



